### ISSUE TITLE
copy_file_range op?

### ISSUE THREAD

**Issue by SUPERCILEX** [axboe/liburing#831]
There's splice but no close_file_range. It'd be great to have.

**Comment by axboe** [comment]
copy_file_range() just uses splice underneath the covers for the generic implementation anyway, can't you just use that directly from io_uring as well?

If not, it could certainly be wired up. But the majority of the work would be enabling nonblocking issue of it, otherwise you'd end up in the same situation potentially as your unlink issue.

**Comment by SUPERCILEX** [comment]
Splice only works if one of the FDs is a pipe according the docs, so that won't work.

**Comment by axboe** [comment]
Right, so you'd splice to a pipe, splice from a pipe. Not ideal, but it'd work.

**Comment by SUPERCILEX** [comment]
Won't that introduce an unnecessary copy?

**Comment by vlovich** [comment]
Yeah, I think a real `copy_file_range` implementation is required. Many filesystems these days support reflinks which `copy_file_range` would take advantage of while splice would generate a copy.

```
       copy_file_range() gives filesystems an opportunity to implement
       "copy acceleration" techniques, such as the use of reflinks
       (i.e., two or more inodes that share pointers to the same copy-
       on-write disk blocks) or server-side-copy (in the case of NFS).
```

So if I'm copying on the same XFS/BTFS/EXT4 filesystem (& recently ZFS too I believe), `copy_file_range` would copy a few inodes max regardless of the file size (+not actually consume that space) whereas `splice` would duplicate the entire file.

**Comment by saltzm** [comment]
FWIW I was also just wishing this existed so that I could make use of reflinks. I think it could also be done using [ioctl_ficlonerange)](https://man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html) but AFAICT there's no way to do that with iouring right now either
